src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java

```diff
@@ -0,0 +1,203 @@
+            return this.ordinal() > other.ordinal();
+        }
+
+        private boolean isLower(final @NotNull State other) {
+            return this.ordinal() < other.ordinal();
+        }
+    }
+
+    private synchronized void advanceState(final @NotNull State newState) {
+        if ((currentState == State.STOPPED && newState == State.NOT_STARTED)
+                || newState.isHigher(currentState)) {
+            log.debug("Advancing EmbeddedHiveMQState from \"{}\" to \"{}\"", currentState, newState);
+            currentState = newState;
+        }
+    }
+
+    @Override
+    public synchronized @NotNull CompletableFuture<Void> start() {
+        advanceState(State.NOT_STARTED);
+        if (currentState == State.NOT_STARTED) {
+            advanceState(State.STARTING);
+            log.info("Starting EmbeddedHiveMQ.");
+
+            startFuture = new CompletableFuture<>();
+            CompletableFuture.runAsync(() -> {
+
+                configurationService = ConfigurationBootstrap.bootstrapConfig(systemInformation);
+
+                bootstrapInjector();
+                final HiveMQServer hiveMQServer = injector.getInstance(HiveMQServer.class);
+                try {
+                    hiveMQServer.start();
+                } catch (final Exception e) {
+                    throw new HiveMQServerException(e);
+                }
+                advanceState(State.RUNNING);
+            }).whenComplete(HiveMQServerException.handler(startFuture));
+        } else {
+            log.warn("Could not start EmbeddedHiveMQ. Reason: \"Already started\"");
+        }
+        return startFuture;
+    }
+
+    private void bootstrapInjector() {
+        if (injector == null) {
+            final HivemqId hiveMQId = new HivemqId();
+            final Injector persistenceInjector =
+                    GuiceBootstrap.persistenceInjector(
+                            systemInformation, metricRegistry, hiveMQId, configurationService);
+
+            try {
+                persistenceInjector.getInstance(PersistenceStartup.class).finish();
+            } catch (final InterruptedException e) {
+                log.error("EmbeddedHiveMQ persistence Startup interrupted.");
+            }
+
+            injector =
+                    GuiceBootstrap.bootstrapInjector(systemInformation, metricRegistry, hiveMQId, configurationService,
+                            persistenceInjector);
+        }
+    }
+
+    @Override
+    public synchronized @NotNull CompletableFuture<Void> stop() {
+        if (currentState == State.NOT_STARTED) {
+            log.warn("Could not stop EmbeddedHiveMQ. Reason: \"Not started\"");
+            stopFuture = CompletableFuture.completedFuture(null);
+
+        } else if (currentState.isLower(State.STOPPING)) {
+            advanceState(State.STOPPING);
+            log.info("Stopping EmbeddedHiveMQ.");
+
+            stopFuture = startFuture.thenRunAsync(() -> {
```

Comment:
I don't think it is a good idea to use a shared executor for stop as well, because this blocks for some time.
I think a new thread here would be the best solution

Reply:
see answer above